### PR TITLE
[BEAM-7909] upgrade python lib versions to match to dataflow worker

### PR DESCRIPTION
**Pull request by Hannah-Jiang**
This PR is upgrading python library versions to match versions installed at dataflow workers. (https://cloud.google.com/dataflow/docs/concepts/sdk-worker-dependencies)
In order to make it easy to track down in case anything is broken, I created a separate PR for this change.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by Hannah-Jiang** [comment]
R: @aaltay

**Comment by Hannah-Jiang** [comment]
Run Portable_Python PreCommit

File: sdks/python/container/base_image_requirements.txt [hunk 1]
```diff
@@ -28,46 +28,46 @@ avro-python3==1.8.2;python_version>="3.4"
 fastavro==0.21.4
 crcmod==1.7
 dill==0.2.9
-future==0.16.0
+future==0.17.1
 futures==3.2.0;python_version<"3.0"
-grpcio==1.15.0
-hdfs==2.1.0
-httplib2==0.9.2
+grpcio==1.21.1
```
Comment (aaltay) [inline sdks/python/container/base_image_requirements.txt]: 1.22.0 ?

File: sdks/python/container/base_image_requirements.txt [hunk 2]
```diff
@@ -28,46 +28,46 @@ avro-python3==1.8.2;python_version>="3.4"
...
+hdfs==2.5.6
+httplib2==0.12.0
 mock==2.0.0
 oauth2client==3.0.0
-protobuf==3.6.1
-pyarrow==0.11.1
-pydot==1.2.4
-pyparsing==2.3.1
-pytz==2018.4
+protobuf==3.8.0
```
Comment (aaltay) [inline sdks/python/container/base_image_requirements.txt]: 3.9.0 ?

File: sdks/python/container/base_image_requirements.txt [hunk 3]
```diff
@@ -28,46 +28,46 @@ avro-python3==1.8.2;python_version>="3.4"
...
+pyarrow==0.13.0
```
Comment (aaltay) [inline sdks/python/container/base_image_requirements.txt]: 0.14.1 ?

File: sdks/python/container/base_image_requirements.txt [hunk 4]
```diff
@@ -28,46 +28,46 @@ avro-python3==1.8.2;python_version>="3.4"
...
+pydot==1.4.1
+pyparsing==2.4.0
```
Comment (aaltay) [inline sdks/python/container/base_image_requirements.txt]: Do you know where this dependency comes from?

File: sdks/python/container/base_image_requirements.txt [hunk 5]
```diff
@@ -28,46 +28,46 @@ avro-python3==1.8.2;python_version>="3.4"
...
-google-cloud-pubsub==0.39.0
+google-apitools==0.5.28
+googledatastore==7.0.2;python_version<"3.0"
+google-cloud-pubsub==0.39.1
 google-cloud-bigquery==1.6.0
 proto-google-cloud-datastore-v1==0.90.4
-google-cloud-bigtable==0.31.1
-google-cloud-core==0.28.1
+google-cloud-bigtable==0.32.1
+google-cloud-core==0.29.1
```
Comment (aaltay) [inline sdks/python/container/base_image_requirements.txt]: 1.0.2 ?

File: sdks/python/container/base_image_requirements.txt [hunk 6]
```diff
@@ -28,46 +28,46 @@ avro-python3==1.8.2;python_version>="3.4"
...
 nose==1.3.7
 
 # GCP extra features
-google-apitools==0.5.26
-googledatastore==7.0.1;python_version<"3.0"
-google-cloud-pubsub==0.39.0
+google-apitools==0.5.28
+googledatastore==7.0.2;python_version<"3.0"
+google-cloud-pubsub==0.39.1
 google-cloud-bigquery==1.6.0
```
Comment (aaltay) [inline sdks/python/container/base_image_requirements.txt]: 1.17.0 (google-cloud-bigquery)

File: sdks/python/container/base_image_requirements.txt [hunk 7]
```diff
@@ -28,46 +28,46 @@ avro-python3==1.8.2;python_version>="3.4"
...
 pyvcf==0.6.8;python_version<"3.0"
-pyyaml==3.12
-typing==3.6.1
+pyyaml==3.13
+typing==3.6.6
 
 # Setup packages
 nose==1.3.7
 
 # GCP extra features
```
Comment (aaltay) [inline sdks/python/container/base_image_requirements.txt]: Also add: google-cloud-datastore==1.7.4

File: sdks/python/container/base_image_requirements.txt [hunk 8]
```diff
@@ -28,46 +28,46 @@ avro-python3==1.8.2;python_version>="3.4"
...
-pandas==0.23.4
-protorpc==0.11.1
+numpy==1.16.4
+scipy==1.2.2
+pandas==0.24.2
+protorpc==0.12.0
 python-gflags==3.0.6
-setuptools<=39.1.0 # requirement for Tensorflow.
-tensorflow==1.11.0
+setuptools==41.0.1 # requirement for Tensorflow.
```
Comment (aaltay) [inline sdks/python/container/base_image_requirements.txt]: I think we can drop this. No need to pin/downgrade it anymore.

File: sdks/python/container/base_image_requirements.txt [hunk 9]
```diff
@@ -28,46 +28,46 @@ avro-python3==1.8.2;python_version>="3.4"
 fastavro==0.21.4
```
Comment (aaltay) [inline sdks/python/container/base_image_requirements.txt]: 0.21.24 ?

Comment (aaltay) [inline sdks/python/container/base_image_requirements.txt]: Check out for other newer dependencies in setup.py. (For example pymongo == 3.8.0 )

File: sdks/python/container/base_image_requirements.txt (hunk 4 above)
Comment (Hannah-Jiang) [reply sdks/python/container/base_image_requirements.txt]: Dependency chain is `pipeline_graph` -> `pydot` -> `dot_parser` -> `pyparsing`.

**Comment by Hannah-Jiang** [comment]
Run Python_PVR_Flink PreCommit

File: sdks/python/container/base_image_requirements.txt (hunk 4 above)
Comment (aaltay) [reply sdks/python/container/base_image_requirements.txt]: We already have pydot in the list. And we do not list other transitive dependencies. Should we remove this from here?

Separately, it is actually a good idea to list all transitive dependencies. There was a mailing list discussion on this. I do not know if there is a JIRA issue about it.

Comment (Hannah-Jiang) [reply sdks/python/container/base_image_requirements.txt]: I removed `pyparsing` from the list.
I haven't found a ticket for listing all transitive dependencies. Could you please explain what it is? I will create a ticket when I have more details.

Comment (aaltay) [reply sdks/python/container/base_image_requirements.txt]: Beam has a list of dependencies (listed in setup.py). These are Beam's direct dependency. Each of these direct dependency will have their own list of dependencies. These are transitive dependencies of Beam. Beam depends on them indirectly and in setup.py Beam does not list them.

In a worker environment it is beneficial to lock version of all Beam dependencies including the indirect ones.

Does it answer your question?

Comment (Hannah-Jiang) [reply sdks/python/container/base_image_requirements.txt]: Yes. Thanks for more details. I created a ticket https://issues.apache.org/jira/browse/BEAM-8084.

**Comment by Hannah-Jiang** [comment]
All tests passed, can we merge it?

**Comment by aaltay** [comment]
Thank you!

@Hannah-Jiang could you also file another JIRA issue for finding a process to keep this list in sync with setup.py.

**Pull request merged**